### PR TITLE
Global model

### PR DESCRIPTION
**Pull request by ykate1998**
# Description

Adding the CNN Image Retrieval model implementation based on the following papers:

[1] Fine-tuning CNN Image Retrieval with No Human Annotation,
Radenović F., Tolias G., Chum O., TPAMI 2018 [arXiv]
https://arxiv.org/abs/1711.02512

[2] CNN Image Retrieval Learns from BoW: Unsupervised Fine-Tuning with Hard
Examples, Radenović F., Tolias G., Chum O., ECCV 2016 [arXiv]
https://arxiv.org/abs/1604.02426

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] A new research paper code implementation

### CODE REVIEW

File: research/delf/delf/python/training/model/global_model.py [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+      global_features_utils.debug_and_log(
+        '\r>>>> {}/{} done...'.format((i + 1), len(images)),
+        debug_on_the_same_line=True)
+  global_features_utils.debug_and_log('', log=False)
+
+  descriptors = tf.transpose(descriptors, perm=[1, 0])
+  return descriptors
+
+
+def extract_ms(net, input, ms, msp):
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/global_model.py:259]: maybe rename to extract_multi_scale_descriptor

File: research/delf/delf/python/training/model/global_model.py [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+  """Extracts the global descriptor multi scale.
+
+  Args:
+    net: Model object, network for the forward pass.
+    input: [B, H, W, C] input tensor in channel-last (BHWC) configuration.
+    ms: List of float scales.
+    msp: Float, multi-scale normalization power parameter.
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/global_model.py:266]: same terminology comment as in the previous function

File: research/delf/delf/python/training/model/global_model_test.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      filename = os.path.join(FLAGS.test_tmpdir, 'test_image_{}.jpg'.format(i))
+      img_out.save(filename)
+      image_paths.append(filename)
+
+    descriptors = global_model.extract_global_descriptors_from_list(
+      model, image_paths, image_size=1024, bounding_boxes=None, ms=[1., 3.],
+      msp=2, print_freq=1)
+    self.assertAllEqual([2048, 2], tf.shape(descriptors))
+
+  def testExtractMS(self):
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/global_model_test.py:72]: testExtractMultiScale

**Review by andrefaraujo — CHANGES_REQUESTED** [review]
Thanks, Kate! It looks good overall, the comments I have are generally minor. Please take a look :)

File: research/delf/delf/python/training/model/global_model.py [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+    else:
+      # Initialize with random weights.
+      net_in = getattr(tf.keras.applications, architecture)(include_top=False,
+                                                            weights=None)
+
+    # Initialize `feature_extractor`. Take only convolutions for
+    # `feature_extractor`, always end with ReLU to make last activations
+    # non-negative.
+    if architecture.lower().startswith('densenet'):
+      net_in.add(tf.keras.layers.ReLU())
```
[the remark this answers is not in the document]
Comment (ykate1998) [reply research/delf/delf/python/training/model/global_model.py:114]: As far as I can tell, all models are imported in keras.Model format, not only EfficientNet. The code you pointed out seems to be working in the case of DenseNet

[the remark this answers is not in the document]
Comment (ykate1998) [reply research/delf/delf/python/training/model/global_model.py:119]: Sorry, you were right, I tested this specific pass and it needs the conversion to Sequential as you mentioned

**Comment by ykate1998** [comment]
@andrefaraujo , I believe it should be fixed now. Thank you for the review

File: research/delf/delf/python/training/model/global_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (andrefaraujo) [reply research/delf/delf/python/training/model/global_model.py:122]: Thanks!

File: research/delf/delf/python/training/model/global_model_test.py [hunk 5]
```diff
@@ -65,19 +65,21 @@ def testExtractVectors(self):
       image_paths.append(filename)
 
     descriptors = global_model.extract_global_descriptors_from_list(
-      model, image_paths, image_size=1024, bounding_boxes=None, ms=[1., 3.],
-      msp=2, print_freq=1)
+            model, image_paths, image_size=1024, bounding_boxes=None,
+            scales=[1., 3.],
+            multi_scale_power=2, print_freq=1)
```
Comment (andrefaraujo) [inline research/delf/delf/python/training/model/global_model_test.py:70]: fix indentation here (pull code to previous line)

**Review by andrefaraujo — CHANGES_REQUESTED** [review]
Thanks Kate! Looks great. Just one small change left, and I will merge the code once you make the edit.

**Comment by ykate1998** [comment]
@andrefaraujo, should be fixed now :) Thank you

**Review by andrefaraujo — APPROVED** [review]
Thanks, Kate!

**Comment by andrefaraujo** [comment]
@ykate1998 can you please resolve the conflict? I cannot merge it right now. Once you resolve the conflict, I'll be able to merge. I guess it may be related to the other PR which I just merged.

**Comment by ykate1998** [comment]
Hello, should be fixed now. Thanks

**Review by andrefaraujo — APPROVED** [review]
Thanks Kate! Merging...

**Pull request merged**